encoder.go

```diff
@@ -29,7 +29,17 @@ func (e *Encoder) WriteField(f HeaderField) error {
 		e.wrotePrefix = true
 	}
 
-	e.writeLiteralFieldWithoutNameReference(f)
+	nameFound, valueFound, idx := e.getHeaderFieldIndex(&f)
```

Comment:
It seems this would be easier if you don't define a `getHeaderFieldIndex` function.
It should also be faster, as you can skip the map lookup for valueFound if nameFound is false.

Reply:
Yeah, initially I thought I'd call it in several places but I ended up just making it harder to read.
Also, while inlining it I also found a bug:
I'm writing an indexed field if nameFound and len(f.Value) == 0 but that would be incorrect for some input.
For example {":method", "" } cannot be encoded with an indexed field.
Will add more test coverage.